modules/renter/hostdb/scan_test.go

```diff
@@ -0,0 +1,156 @@
+package hostdb
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// TestDecrementReliability tests the decrementReliability method.
+func TestDecrementReliability(t *testing.T) {
+	hdb := bareHostDB()
+
+	// Decrementing a non-existent host should be a no-op.
+	// NOTE: can't check any post-conditions here; only indication of correct
+	// behavior is that the test doesn't panic.
+	hdb.decrementReliability("foo", types.NewCurrency64(0))
+
+	// Add a host to allHosts and activeHosts. Decrementing it should remove it
+	// from activeHosts.
+	h := new(hostEntry)
+	h.NetAddress = "foo"
+	h.reliability = types.NewCurrency64(1)
+	hdb.allHosts[h.NetAddress] = h
+	hdb.activeHosts[h.NetAddress] = &hostNode{hostEntry: h}
+	hdb.decrementReliability(h.NetAddress, types.NewCurrency64(0))
+	if _, ok := hdb.activeHosts[h.NetAddress]; ok {
+		t.Error("decrementing did not remove host from activeHosts")
+	}
+
+	// Decrement reliability to 0. This should remove the host from allHosts.
+	hdb.decrementReliability(h.NetAddress, h.reliability)
+	if _, ok := hdb.allHosts[h.NetAddress]; ok {
+		t.Error("decrementing did not remove host from allHosts")
+	}
+}
+
+// probeDialer is used to test the threadedProbeHosts method. A simple type
+// alias is used so that it can easily be redefined during testing, allowing
+// multiple behaviors to be tested.
+type probeDialer func(modules.NetAddress, time.Duration) (net.Conn, error)
+
+func (dial probeDialer) DialTimeout(addr modules.NetAddress, timeout time.Duration) (net.Conn, error) {
+	return dial(addr, timeout)
+}
+
+// TestThreadedProbeHosts tests the threadedProbeHosts method.
+func TestThreadedProbeHosts(t *testing.T) {
+	hdb := bareHostDB()
+
+	// create a host to send to threadedProbeHosts
+	h := new(hostEntry)
+	h.NetAddress = "foo"
+	h.reliability = baseWeight // enough to withstand a few failures
+
+	// define a helper function for running threadedProbeHosts. We send the
+	// hostEntry, close the channel, and then call threadedProbeHosts.
+	// threadedProbeHosts will receive the host, loop once, and return after
+	// seeing the channel has closed.
+	runProbe := func(h *hostEntry) {
+		hdb.scanPool <- h
+		close(hdb.scanPool)
+		hdb.threadedProbeHosts()
+		// reset hdb.scanPool
+		hdb.scanPool = make(chan *hostEntry, 1)
+	}
+
+	// make the dial fail
+	hdb.dialer = probeDialer(func(modules.NetAddress, time.Duration) (net.Conn, error) {
+		return nil, net.UnknownNetworkError("fail")
+	})
+	runProbe(h)
+	if _, ok := hdb.activeHosts[h.NetAddress]; ok {
```

Comment:
you can probably use an exported function here to count the number of hosts, instead of directly looking at the memory. This helps to keep the test behavior-oriented instead of implementation-oriented.